src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs

```diff
@@ -1,15 +1,110 @@
+			return visitor.Visit(expressionTree);
+		}
+
+		// _partialEvaluationInfo contains a list of the expressions that are safe to be evaluated.
+		private readonly PartialEvaluationInfo _partialEvaluationInfo;
+		private readonly IEvaluatableExpressionFilter _evaluatableExpressionFilter;
+
+		private NhPartialEvaluatingExpressionVisitor(
+			PartialEvaluationInfo partialEvaluationInfo,
+			IEvaluatableExpressionFilter evaluatableExpressionFilter)
+		{
+			_partialEvaluationInfo = partialEvaluationInfo ?? throw new ArgumentNullException(nameof(partialEvaluationInfo));
+			_evaluatableExpressionFilter = evaluatableExpressionFilter ?? throw new ArgumentNullException(nameof(evaluatableExpressionFilter));
+		}
+
+		public override Expression Visit(Expression expression)
+		{
+			// Only evaluate expressions which do not use any of the surrounding parameter expressions. Don't evaluate
+			// lambda expressions (even if you could), we want to analyze those later on.
+			if (expression == null)
+				return null;
+
+			if (expression.NodeType == ExpressionType.Lambda || !_partialEvaluationInfo.IsEvaluatableExpression(expression))
+				return base.Visit(expression);
+
+			Expression evaluatedExpression;
+			try
+			{
+				evaluatedExpression = EvaluateSubtree(expression);
+			}
+			catch (Exception ex)
+			{
+				// Evaluation caused an exception. Skip evaluation of this expression and proceed as if it weren't evaluable.
+				var baseVisitedExpression = base.Visit(expression);
+				// Then wrap the result to capture the exception for the back-end.
+				return new PartialEvaluationExceptionExpression(ex, baseVisitedExpression);
+			}
+
+			if (evaluatedExpression != expression)
+				return EvaluateIndependentSubtrees(evaluatedExpression, _evaluatableExpressionFilter);
+
+			return evaluatedExpression;
+		}
+
+		/// <summary>
+		/// Evaluates an evaluatable <see cref="Expression"/> subtree, i.e. an independent expression tree that is compilable and executable
+		/// without any data being passed in. The result of the evaluation is returned as a <see cref="ConstantExpression"/>; if the subtree
+		/// is already a <see cref="ConstantExpression"/>, no evaluation is performed.
+		/// </summary>
+		/// <param name="subtree">The subtree to be evaluated.</param>
+		/// <returns>A <see cref="ConstantExpression"/> holding the result of the evaluation.</returns>
+		private Expression EvaluateSubtree(Expression subtree)
+		{
+			if (subtree == null) throw new ArgumentNullException(nameof(subtree));
+
+			if (subtree.NodeType == ExpressionType.Constant)
+			{
+				var constantExpression = (ConstantExpression) subtree;
+				var valueAsIQueryable = constantExpression.Value as IQueryable;
+				if (valueAsIQueryable != null && !IsEvaluatedQueryable(constantExpression.Value) && valueAsIQueryable.Expression != constantExpression)
+					return valueAsIQueryable.Expression;
```

Comment:
I guess this is here that code differs. It seems to create a bunch of test failures, but I have not tried analyzing them yet.

Reply:
I'll have a look.